lib/attributes.js

```diff
@@ -13,162 +13,283 @@
-   var frameControl = (d[1]<< 8) + d[0];
-   var productId = (d[3]<<8) + d[2];
-   var counter = d[4];
-   var r = {
-    frameControl : frameControl,
-    productId : productId,
-    counter : counter
-   };
-   switch (productId) {
-     case 0x5d: r.productName = "HHCCPOT002"; break;
-     case 0x98: r.productName = "HHCCJCY01"; break;
-     case 0x1d8: r.productName = "Stratos"; break;
-     case 0x2df: r.productName = "JQJCY01YM"; break;
-     case 0x3b6: r.productName = "YLKG08YL"; break;
-     case 0x3bc: r.productName = "GCLS002"; break;
-     case 0x40a: r.productName = "WX08ZM"; break;
-     case 0x45b: r.productName = "LYWSD02"; break;
-     case 0x55b: r.productName = "LYWSD03MMC"; break;
-     case 0x576: r.productName = "CGD1"; break;
-     case 0x347: r.productName = "CGG1"; break;
-     case 0x1aa: r.productName = "LYWSDCGQ"; break;
-   }
-
-   if (d[11] == 0x04) {
-     var temp = d[15] << 8 | d[14];
-     if (temp & 0x8000) temp -= 0x10000;
-     r.temp = temp / 10;
-   } else if (d[11] == 0x06) {
-     r.humidity = (d[15] << 8 | d[14]) / 10;
-   } else if (d[11] == 0x0a) {
-     r.battery = d[14];
-   } else if (d[11] == 0x0d) {
-     var temp = d[15] << 8 | d[14];
-     if (temp & 0x8000) temp -= 0x10000;
-     r.temp = temp / 10;
-     r.humidity = (d[17] << 8 | d[16]) / 10;
-   }
-   return r;
- },
- "feaa" : function(d) { // Eddystone
-   if (d[0]==0x10) { // URL
-     var rssi = d[1];
-     if (rssi&128) rssi-=256; // signed number
-     var urlType = d[2];
-     var URL_TYPES = [
+  "1809": function (a) { // Temperature
+    var t = (a.length == 2) ? (((a[1] << 8) + a[0]) / 100) : a[0];
+    if (t >= 128) t -= 256;
+    return {temp: t}
+  },
+  "180f": function (a) { // Battery percent
+    return {
+      battery: a[0]
+    }
+  },
+  "181a": function (a) { // ATC_MiThermometer
+    if (a.length >= 15) { // pvvx
+      let voltage = a.readUInt16LE(10);
+      return {
+        temp: a.readUInt16LE(6) / 100,
+        humidity: a.readUInt16LE(8) / 100,
+        battery_voltage: voltage > 1000 ? voltage / 1000 : voltage,
+        battery: a.readUInt8(12),
+        counter: a.readUInt8(13),
+        flg: a.readUInt8(14)
+      }
+    } else if (a.length === 13) {
+      return {
+        temp: (a[7] | (a[6] << 8)) / 10,
+        humidity: a[8],
+        battery: a[9],
+        battery_voltage: (a[11] | (a[10] << 8)) / 1000
+      }
+    }
+  },
+  "181b": function (a) { // Xiaomi V2 Scale
+    let unit;
+    let weight = a.readUInt16LE(a.length - 2) / 100;
+    if ((a[0] & (1 << 4)) !== 0) { // Chinese Catty
+      unit = "jin";
+    } else if ((a[0] & 0x0F) === 0x03) { // Imperial pound
+      unit = "lbs";
+    } else if ((a[0] & 0x0F) === 0x02) { // MKS kg
+      unit   = "kg";
+      weight = weight / 2;
+    } else {
+      unit = "???"
+    }
+    const state = {
+      isStabilized: ((a[1] & (1 << 5)) === 1),
+      loadRemoved: ((a[1] & (1 << 7)) === 1),
+      impedanceMeasured: ((a[1] & (1 << 1)) === 1)
```

Comment:
I'm afraid this code is broken. These statements will *never* be true. The original code you had (`!== 0`) would have worked fine though.

Reply:
fixed